### PR TITLE
Initial implementation of Applications API

### PR DESCRIPTION
**Pull request by malinthaprasan**

### CODE REVIEW

File: components/apimgt/org.wso2.carbon.apimgt.rest.api/src/main/java/org/wso2/carbon/apimgt/rest/api/impl/ApplicationsApiServiceImpl.java [hunk 1]
```diff
@@ -15,29 +25,132 @@
...
+        List<ApplicationDTO> applicationDTOList = new ArrayList<>();
+        try {
+            APIConsumer apiConsumer = APIManagerFactory.getInstance().getAPIConsumer(username);
+            Application[] applications = apiConsumer.getApplications(new Subscriber(subscriber), groupId);
+            for (Application application : applications) {
+                applicationDTOList.add(ApplicationMappingUtil.fromApplicationtoDTO(application));
+            }
+            return Response.ok().entity(applicationDTOList).build();
+        } catch (APIManagementException e) {
+            throw new InternalServerErrorException(e);
```
Comment (nuwand) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api/src/main/java/org/wso2/carbon/apimgt/rest/api/impl/ApplicationsApiServiceImpl.java]: Use proper error handling.

File: components/apimgt/org.wso2.carbon.apimgt.rest.api/src/main/java/org/wso2/carbon/apimgt/rest/api/impl/ApplicationsApiServiceImpl.java [hunk 2]
```diff
@@ -15,29 +25,132 @@
...
+            subscriber = username;
+        }
+
+        try {
+            APIConsumer apiConsumer = APIManagerFactory.getInstance().getAPIConsumer(username);
+            Application application = ApplicationMappingUtil.fromDTOtoApplication(body, new Subscriber(subscriber));
+            String status = apiConsumer.addApplication(application, subscriber);  //todo: use "status" properly
+            return Response.ok().entity(new ApiResponseMessage(ApiResponseMessage.OK, status)).build();
+        } catch (APIManagementException e) {
+            throw new InternalServerErrorException(e);
```
Comment (nuwand) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api/src/main/java/org/wso2/carbon/apimgt/rest/api/impl/ApplicationsApiServiceImpl.java]: Need proper exception handling.

File: components/apimgt/org.wso2.carbon.apimgt.rest.api/src/main/java/org/wso2/carbon/apimgt/rest/api/impl/ApplicationsApiServiceImpl.java (hunk 1 above)
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api/src/main/java/org/wso2/carbon/apimgt/rest/api/impl/ApplicationsApiServiceImpl.java]: Fixed in #1303 (https://github.com/wso2/carbon-apimgt/commit/10b904553097e950f881465d125ef962ac32a3a3)

File: components/apimgt/org.wso2.carbon.apimgt.rest.api/src/main/java/org/wso2/carbon/apimgt/rest/api/impl/ApplicationsApiServiceImpl.java (hunk 2 above)
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api/src/main/java/org/wso2/carbon/apimgt/rest/api/impl/ApplicationsApiServiceImpl.java]: Fixed in #1303 (https://github.com/wso2/carbon-apimgt/commit/10b904553097e950f881465d125ef962ac32a3a3)

File: components/apimgt/org.wso2.carbon.apimgt.rest.api/src/main/java/org/wso2/carbon/apimgt/rest/api/impl/ApplicationsApiServiceImpl.java [hunk 3]
```diff
@@ -15,29 +25,132 @@
...
+                    applicationDTO = ApplicationMappingUtil.fromApplicationtoDTO(application);
+                }
+            }
+            if (applicationDTO != null) {
+                return Response.ok().entity(applicationDTO).build();
+            } else {
+                throw new NotFoundException();
+            }
+        } catch (APIManagementException e) {
+            throw new InternalServerErrorException(e);
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api/src/main/java/org/wso2/carbon/apimgt/rest/api/impl/ApplicationsApiServiceImpl.java]: Already fixed error handling in #1303 (https://github.com/wso2/carbon-apimgt/commit/10b904553097e950f881465d125ef962ac32a3a3)

File: components/apimgt/org.wso2.carbon.apimgt.rest.api/src/main/java/org/wso2/carbon/apimgt/rest/api/impl/ApplicationsApiServiceImpl.java [hunk 4]
```diff
@@ -15,29 +25,132 @@
...
+            subscriber = username;
+        }
+        body.setApplicationId(Integer.parseInt(applicationId));
+        Application application = ApplicationMappingUtil.fromDTOtoApplication(body, new Subscriber(subscriber));
+        try {
+            APIConsumer apiConsumer = APIManagerFactory.getInstance().getAPIConsumer(username);
+            apiConsumer.updateApplication(application);
+            return Response.ok().build();
+        } catch (APIManagementException e) {
+            throw new InternalServerErrorException(e);
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api/src/main/java/org/wso2/carbon/apimgt/rest/api/impl/ApplicationsApiServiceImpl.java]: Already fixed error handling in #1303 (https://github.com/wso2/carbon-apimgt/commit/10b904553097e950f881465d125ef962ac32a3a3)

File: components/apimgt/org.wso2.carbon.apimgt.rest.api/src/main/java/org/wso2/carbon/apimgt/rest/api/impl/ApplicationsApiServiceImpl.java [hunk 5]
```diff
@@ -15,29 +25,132 @@
...
+                    break;
+                }
+            }
+            if (removed) {
+                return Response.ok().build();
+            } else {
+                throw new NotFoundException();
+            }
+        } catch (APIManagementException e) {
+            throw new InternalServerErrorException(e);
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api/src/main/java/org/wso2/carbon/apimgt/rest/api/impl/ApplicationsApiServiceImpl.java]: Already fixed error handling in #1303 (https://github.com/wso2/carbon-apimgt/commit/10b904553097e950f881465d125ef962ac32a3a3)

**Pull request merged**